File: server/services/file_processor.py
import os
from pathlib import Path
from typing import Dict, List
import csv
from io import StringIO
from pypdf import PdfReader


class FileProcessor:
    """Process documents and create text chunks"""

    def __init__(self, chunk_size: int = 1500, chunk_overlap: int = 300):
        self.chunk_size = chunk_size
        self.chunk_overlap = chunk_overlap
        self.max_file_size = 50 * 1024 * 1024  # 50MB

    def validate_file(self, file_path: str) -> Dict:
        """Validate file exists, size, and type"""
        try:
            path = Path(file_path)

            if not path.exists():
                return {"valid": False, "error": "File does not exist"}

            file_size = path.stat().st_size

            if file_size > self.max_file_size:
                size_mb = file_size / (1024 * 1024)
                return {
                    "valid": False,
                    "error": f"File too large: {size_mb:.2f}MB (max 50MB)"
                }

            ext = path.suffix.lower()
            if ext not in ['.pdf', '.csv', '.txt']:
                return {
                    "valid": False,
                    "error": f"Unsupported file type: {ext}"
                }

            return {
                "valid": True,
                "size": file_size,
                "type": ext
            }

        except Exception as e:
            return {"valid": False, "error": str(e)}

    def create_chunks(self, text: str) -> List[Dict]:
        """Create overlapping chunks from text"""
        chunks = []
        chunk_index = 0
        start_pos = 0

        while start_pos < len(text):
            end_pos = start_pos + self.chunk_size

            # Get chunk text
            chunk_text = text[start_pos:end_pos]

            # Try to end at sentence boundary if not at end
            if end_pos < len(text):
                # Look for sentence endings
                last_period = chunk_text.rfind('.')
                last_newline = chunk_text.rfind('\n')
                boundary_index = max(last_period, last_newline)

                # If we found a good boundary in the second half of chunk
                if boundary_index > self.chunk_size * 0.5:
                    chunk_text = chunk_text[:boundary_index + 1]
                    end_pos = start_pos + boundary_index + 1

            chunks.append({
                "text": chunk_text.strip(),
                "index": chunk_index,
                "startChar": start_pos,
                "endChar": end_pos
            })

            chunk_index += 1
            start_pos = end_pos - self.chunk_overlap

        return chunks

    def process_pdf_streaming(self, file_path: str) -> Dict:
        """
        Process PDF in streaming mode to avoid memory issues
        Creates chunks incrementally as pages are processed
        """
        try:
            print(f"Processing PDF in streaming mode: {Path(file_path).name}")

            reader = PdfReader(file_path, strict=False)  # Use non-strict mode to handle malformed PDFs
            total_pages = len(reader.pages)

            print(f"PDF has {total_pages} pages. Processing incrementally...")

            chunks = []
            accumulated_text = ""
            chunk_index = 0
            total_chars = 0

            # Process pages one by one
            for page_num in range(total_pages):
                try:
                    page = reader.pages[page_num]
                    page_text = page.extract_text() or ""  # Handle None return
                    total_chars += len(page_text)

                    accumulated_text += page_text + " "
                except Exception as page_error:
                    print(f"Warning: Failed to extract page {page_num + 1}: {page_error}")
                    # Continue with next page
                    continue

                # When we have enough text, create chunks
                while len(accumulated_text) >= self.chunk_size:
                    chunk = accumulated_text[:self.chunk_size]

                    # Try to end at sentence boundary
                    last_period = chunk.rfind('.')
                    last_newline = chunk.rfind('\n')
                    boundary_index = max(last_period, last_newline)

                    final_chunk = chunk
                    consumed_length = self.chunk_size

                    if boundary_index > self.chunk_size * 0.5:
                        final_chunk = chunk[:boundary_index + 1]
                        consumed_length = boundary_index + 1

                    chunks.append({
                        "text": final_chunk.strip(),
                        "index": chunk_index,
                        "startChar": total_chars - len(accumulated_text),
                        "endChar": total_chars - len(accumulated_text) + consumed_length
                    })

                    chunk_index += 1

                    # Keep overlap for next chunk
                    accumulated_text = accumulated_text[consumed_length - self.chunk_overlap:]

                # Log progress
                if (page_num + 1) % 10 == 0 or page_num + 1 == total_pages:
                    print(f"Processed {page_num + 1}/{total_pages} pages, created {len(chunks)} chunks so far")

            # Process remaining accumulated text
            if accumulated_text.strip():
                chunks.append({
                    "text": accumulated_text.strip(),
                    "index": chunk_index,
                    "startChar": total_chars - len(accumulated_text),
                    "endChar": total_chars
                })

            print(f"PDF streaming complete: {total_pages} pages, {total_chars} characters, {len(chunks)} chunks")

            return {
                "chunks": chunks,
                "metadata": {
                    "pageCount": total_pages,
                    "characterCount": total_chars,
                    "chunkCount": len(chunks)
                }
            }

        except Exception as e:
            print(f"PDF streaming error: {e}")
            raise Exception(f"Failed to stream PDF: {str(e)}")

    def process_csv(self, file_path: str) -> Dict:
        """Process CSV file and create chunks"""
        try:
            print(f"Processing CSV: {Path(file_path).name}")

            with open(file_path, 'r', encoding='utf-8') as f:
                csv_content = f.read()

            # Parse CSV
            csv_reader = csv.DictReader(StringIO(csv_content))
            rows = list(csv_reader)

            # Convert to text format
            text_parts = []
            for row in rows:
                row_text = ", ".join([f"{k}: {v}" for k, v in row.items()])
                text_parts.append(row_text)

            full_text = "\n".join(text_parts)
            chunks = self.create_chunks(full_text)

            print(f"CSV processing complete: {len(rows)} rows, {len(full_text)} characters, {len(chunks)} chunks")

            return {
                "chunks": chunks,
                "metadata": {
                    "rowCount": len(rows),
                    "columnCount": len(rows[0].keys()) if rows else 0,
                    "characterCount": len(full_text),
                    "chunkCount": len(chunks)
                }
            }

        except Exception as e:
            print(f"CSV processing error: {e}")
            raise Exception(f"Failed to process CSV: {str(e)}")

    def process_text(self, file_path: str) -> Dict:
        """Process plain text file and create chunks"""
        try:
            print(f"Processing text file: {Path(file_path).name}")

            with open(file_path, 'r', encoding='utf-8') as f:
                text = f.read()

            chunks = self.create_chunks(text)

            # Count lines and words
            lines = text.split('\n')
            words = text.split()

            print(f"Text processing complete: {len(lines)} lines, {len(words)} words, {len(chunks)} chunks")

            return {
                "chunks": chunks,
                "metadata": {
                    "lineCount": len(lines),
                    "wordCount": len(words),
                    "characterCount": len(text),
                    "chunkCount": len(chunks)
                }
            }

        except Exception as e:
            print(f"Text processing error: {e}")
            raise Exception(f"Failed to process text: {str(e)}")

File: src/services/file-processor.ts
import fs from 'fs/promises';
import path from 'path';
// import pdf from 'pdf-parse'; // Disabled for browser build
// import { parse } from 'csv-parse/sync'; // Disabled for browser build

class FileProcessor {
  private supportedTypes: string[];
  private chunkSize: number;
  private chunkOverlap: number;

  constructor() {
    this.supportedTypes = ['.pdf', '.csv', '.txt'];
    this.chunkSize = 1000; // characters per chunk
    this.chunkOverlap = 200; // overlap between chunks
  }

  /**
   * Check if file type is supported
   */
  isSupportedFile(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return this.supportedTypes.includes(ext);
  }

  /**
   * Process a file and extract text content
   */
  async processFile(filePath) {
    try {
      const ext = path.extname(filePath).toLowerCase();
      const fileName = path.basename(filePath);

      console.log(`Processing file: ${fileName}`);

      let text = '';
      const metadata = {
        fileName,
        filePath,
        fileType: ext,
        processedAt: new Date().toISOString(),
      };

      switch (ext) {
        case '.pdf':
          text = await this.processPDF(filePath);
          break;
        case '.csv':
          text = await this.processCSV(filePath);
          break;
        case '.txt':
          text = await this.processTXT(filePath);
          break;
        default:
          throw new Error(`Unsupported file type: ${ext}`);
      }

      metadata.characterCount = text.length;
      metadata.wordCount = text
        .split(/\s+/)
        .filter((word) => word.length > 0).length;

      return {
        success: true,
        text,
        metadata,
      };
    } catch (error) {
      console.error('Error processing file:', error);
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Process PDF file with true streaming - process page by page without
   * accumulating all text in memory
   */
  async processPDF(filePath) {
    try {
      console.log(`Processing PDF: ${path.basename(filePath)}`);

      // Read file buffer
      const dataBuffer = await fs.readFile(filePath);
      console.log(
        `PDF buffer size: ${(dataBuffer.length / 1024 / 1024).toFixed(2)} MB`
      );

      // Parse PDF to get page count first (without extracting all text)
      const pdfInfo = await pdf(dataBuffer, { max: 1 });
      const totalPages = pdfInfo.numpages;

      console.log(`PDF has ${totalPages} pages. Processing in batches...`);

      // Process pages in small batches to avoid memory buildup
      const batchSize = 5; // Process 5 pages at a time
      let allText = '';

      for (let startPage = 1; startPage <= totalPages; startPage += batchSize) {
        const endPage = Math.min(startPage + batchSize - 1, totalPages);

        // Process this batch of pages
        const batchOptions = {
          max: endPage,
          pagerender: async function (pageData) {
            const pageNum = pageData.pageIndex + 1;
            if (pageNum < startPage) return ''; // Skip pages before this batch

            const textContent = pageData.getTextContent();
            const text = (await textContent).items
              .map((item) => item.str)
              .join(' ');
            return text;
          },
        };

        const batchData = await pdf(dataBuffer, batchOptions);
        allText += batchData.text + '\n\n';

        // Log progress
        console.log(`Processed pages ${startPage}-${endPage}/${totalPages}`);

        // Give garbage collector a chance to run
        if (global.gc) {
          global.gc();
        }
      }

      console.log(
        `PDF processing complete: ${totalPages} pages, ${allText.length} characters`
      );

      return allText.trim();
    } catch (error) {
      console.error('PDF processing error:', error);
      throw new Error(`Failed to parse PDF: ${error.message}`);
    }
  }

  /**
   * Process CSV file
   */
  async processCSV(filePath) {
    try {
      const fileContent = await fs.readFile(filePath, 'utf-8');

      // Parse CSV
      const records = parse(fileContent, {
        columns: true,
        skip_empty_lines: true,
        trim: true,
      });

      // Convert records to text format
      // Format: "Column1: value1, Column2: value2, ..."
      const textLines = records.map((record, index) => {
        const line = Object.entries(record)
          .map(([key, value]) => `${key}: ${value}`)
          .join(', ');
        return `Row ${index + 1}: ${line}`;
      });

      return textLines.join('\n');
    } catch (error) {
      throw new Error(`Failed to parse CSV: ${error.message}`);
    }
  }

  /**
   * Process TXT file
   */
  async processTXT(filePath) {
    try {
      return await fs.readFile(filePath, 'utf-8');
    } catch (error) {
      throw new Error(`Failed to read TXT file: ${error.message}`);
    }
  }

  /**
   * Validate file before processing
   */
  async validateFile(filePath) {
    try {
      // Check if file exists
      await fs.access(filePath);

      // Check file size (max 50MB)
      const stats = await fs.stat(filePath);
      const fileSizeInMB = stats.size / (1024 * 1024);

      if (fileSizeInMB > 50) {
        return {
          valid: false,
          error: `File is too large (${fileSizeInMB.toFixed(2)}MB). Maximum size is 50MB.`,
        };
      }

      // Check file type
      if (!this.isSupportedFile(filePath)) {
        const ext = path.extname(filePath);
        return {
          valid: false,
          error: `Unsupported file type: ${ext}. Supported types: ${this.supportedTypes.join(', ')}`,
        };
      }

      return {
        valid: true,
        size: stats.size,
        sizeInMB: fileSizeInMB,
      };
    } catch (error) {
      return {
        valid: false,
        error: `File validation failed: ${error.message}`,
      };
    }
  }

  /**
   * Get file information
   */
  async getFileInfo(filePath) {
    try {
      const stats = await fs.stat(filePath);
      const ext = path.extname(filePath).toLowerCase();
      const name = path.basename(filePath);

      return {
        name,
        path: filePath,
        size: stats.size,
        sizeInMB: (stats.size / (1024 * 1024)).toFixed(2),
        type: ext,
        supported: this.isSupportedFile(filePath),
        modifiedAt: stats.mtime,
      };
    } catch (error) {
      throw new Error(`Failed to get file info: ${error.message}`);
    }
  }

  /**
   * Split text into overlapping chunks for better context preservation
   */
  chunkText(text) {
    const chunks = [];
    let startIndex = 0;

    while (startIndex < text.length) {
      // Get chunk with specified size
      const endIndex = Math.min(startIndex + this.chunkSize, text.length);
      let chunk = text.substring(startIndex, endIndex);

      // Try to end chunk at a sentence boundary if possible
      if (endIndex < text.length) {
        const lastPeriod = chunk.lastIndexOf('.');
        const lastNewline = chunk.lastIndexOf('\n');
        const boundaryIndex = Math.max(lastPeriod, lastNewline);

        if (boundaryIndex > this.chunkSize * 0.5) {
          // Only use boundary if it's in the latter half of the chunk
          chunk = chunk.substring(0, boundaryIndex + 1);
        }
      }

      chunks.push({
        text: chunk.trim(),
        index: chunks.length,
        startChar: startIndex,
        endChar: startIndex + chunk.length,
      });

      // Move start index forward, accounting for overlap
      startIndex += chunk.length - this.chunkOverlap;
    }

    return chunks;
  }

  /**
   * Process file and return with chunks
   */
  async processFileWithChunks(filePath) {
    try {
      const result = await this.processFile(filePath);

      if (!result.success) {
        return result;
      }

      // Chunk the text
      const chunks = this.chunkText(result.text);

      // Clear the full text from memory to save space
      delete result.text;

      return {
        ...result,
        chunks,
        metadata: {
          ...result.metadata,
          chunkCount: chunks.length,
          chunkSize: this.chunkSize,
          chunkOverlap: this.chunkOverlap,
        },
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
      };
    }
  }

  /**
   * Process PDF in streaming mode - chunk as we go to minimize memory usage
   * Returns chunks directly without keeping full text in memory
   */
  async processPDFStreaming(filePath) {
    try {
      console.log(
        `Processing PDF in streaming mode: ${path.basename(filePath)}`
      );

      const dataBuffer = await fs.readFile(filePath);
      console.log(
        `PDF buffer size: ${(dataBuffer.length / 1024 / 1024).toFixed(2)} MB`
      );

      // Get page count
      const pdfInfo = await pdf(dataBuffer, { max: 1 });
      const totalPages = pdfInfo.numpages;

      console.log(
        `PDF has ${totalPages} pages. Processing and chunking incrementally...`
      );

      const chunks = [];
      let accumulatedText = '';
      let chunkIndex = 0;
      let totalChars = 0;

      // Process pages one by one
      for (let pageNum = 1; pageNum <= totalPages; pageNum++) {
        const pageOptions = {
          max: pageNum,
          pagerender: async function (pageData) {
            if (pageData.pageIndex + 1 < pageNum) return '';
            const textContent = pageData.getTextContent();
            const text = (await textContent).items
              .map((item) => item.str)
              .join(' ');
            return text;
          },
        };

        const pageData = await pdf(dataBuffer, pageOptions);
        const pageText = pageData.text;
        totalChars += pageText.length;

        accumulatedText += pageText + ' ';

        // When we have enough text, create chunks
        while (accumulatedText.length >= this.chunkSize) {
          const chunk = accumulatedText.substring(0, this.chunkSize);

          // Try to end at sentence boundary
          const lastPeriod = chunk.lastIndexOf('.');
          const lastNewline = chunk.lastIndexOf('\n');
          const boundaryIndex = Math.max(lastPeriod, lastNewline);

          let finalChunk = chunk;
          let consumedLength = this.chunkSize;

          if (boundaryIndex > this.chunkSize * 0.5) {
            finalChunk = chunk.substring(0, boundaryIndex + 1);
            consumedLength = boundaryIndex + 1;
          }

          chunks.push({
            text: finalChunk.trim(),
            index: chunkIndex++,
            startChar: totalChars - accumulatedText.length,
            endChar: totalChars - accumulatedText.length + consumedLength,
          });

          // Keep overlap for next chunk
          accumulatedText = accumulatedText.substring(
            consumedLength - this.chunkOverlap
          );
        }

        // Log progress
        if (pageNum % 10 === 0 || pageNum === totalPages) {
          console.log(
            `Processed ${pageNum}/${totalPages} pages, created ${chunks.length} chunks so far`
          );
        }

        // Force garbage collection if available
        if (global.gc && pageNum % 10 === 0) {
          global.gc();
        }
      }

      // Process remaining accumulated text
      if (accumulatedText.trim().length > 0) {
        chunks.push({
          text: accumulatedText.trim(),
          index: chunkIndex++,
          startChar: totalChars - accumulatedText.length,
          endChar: totalChars,
        });
      }

      console.log(
        `PDF streaming complete: ${totalPages} pages, ${totalChars} characters, ${chunks.length} chunks`
      );

      return {
        chunks,
        metadata: {
          pageCount: totalPages,
          characterCount: totalChars,
          chunkCount: chunks.length,
        },
      };
    } catch (error) {
      console.error('PDF streaming error:', error);
      throw new Error(`Failed to stream PDF: ${error.message}`);
    }
  }
}

// Export singleton instance
export default new FileProcessor();
